### ISSUE TITLE
$set on Map keys fails in TypeScript when typed, otherwise runs properly

### ISSUE THREAD

**Issue by code-hunger** [kolodny/immutability-helper#120]
I wanted to show a simpler example to the comment I made in #103 

    var simpleState = new Map<string, number>([["a", 0]]);

    var updatedState = update(simpleState, { a: { $set: 1 } });

    expect(updatedState).toEqual(new Map([["a", 1]]));

This passes, as long as `simpleState` is cast to `any` before being `update`d.

Here's the type error when no such cast is made:

*Argument of type '{ a: { $set: number; }; }' is not assignable to parameter of type 'Spec<Map<string, number>, never>'.
  Object literal may only specify known properties, and 'a' does not exist in type 'Spec<Map<string, number>, never>'*

I'll paste my thoughts on this from my other comment:

> I cannot imagine if and how hard it could be to implement this in TS, because it has to infer that the type of the key `['mapKey']` and then decide whether it matches the first template argument of `Map<KeyT, ValueT>`, a `string` by which the properties of the Map object itself are accessed (e.g. `map.size`), or neither.  
> E.g., when `KeyT` is string, and you are `$set`ting a `"size"`, it is ambiguous whether you want to do `map.set('size', x)`, or `map.size = x`. It is clear you shouldn't be allowed to do the second one, but what about other properties of map?
> 
> If you do `["something"]: { $set: value }`, then what will happen? If we take the logical way, it should `map.set('something', value)`, but that onlly works for KeyT being a string. In other cases it should choose the second way, but this introduces inconsistency which might be undesirable. And if you do not, we're back to the previous problem with native Map props, with the added inconvenience of being unable to call `map.set()` with the `$set` shortcut at all.

**Comment by lixiaoyang1992** [comment]
same problem

**Comment by zaguiini** [comment]
Same, but only on `$add` instead of `$set`.

**Comment by wulucxy** [comment]
same problem here。

@code-hunger any progress?

**Comment by code-hunger** [comment]
As I said, the problem is that the code is ambiguous, and the type checker doesn't know if we want to call Map::set() or set a regular property on the map object.

I don't see a clear solution to the problem. One possibility is to restrict `update()`s on Map to only calling `.set()` and not allow modifications on the map object, which makes update() inconsistent.

The inherent problem is that `update()` is meant to update pure JS objects/arrays, and not concrete Map/Set/... instances - we're kind of 'abusing' it to do so. As we see, this inevitably leads to problems [noise, rather], which is the price for misusing the tool.

**Comment by kolodny** [comment]
It looks like the `MapSpec` needs to change to

```ts
declare type MapSpec<K, V> = {
    $add: Array<[K, V]>;
} | {
    $remove: K[];
} | {
  [key: string]: { $set: V }
} ;
```

Does anyone want to create a PR with tests for this? SetSpec may need to change similarly as well (I'm unsure about that though)

**Comment by kolodny** [comment]
Done in c60fdba and published to `v3.0.1`
